File: ruby/muse/data_structures/ArrayList.rb
module ArrayList
  class ArrayList
    @@DEFAULT_CAPACITY = 64

    def initialize(physicalSize = 0)
      @data = nil
      @logicalSize = 0
      @physicalSize = @@DEFAULT_CAPACITY

      @physicalSize = physicalSize if physicalSize > 1
      @data = Array.new(@physicalSize)
    end

    def size
      @logicalSize
    end

    def isEmpty
      @logicalSize.zero?
    end

    def get(index)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index >= @logicalSize

      @data[index]
    end

    def set(index, element)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index >= @logicalSize

      @data[index] = element
    end

    def insert(index, element)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index > @logicalSize

      if @logicalSize == @physicalSize
        newCapacity = @@DEFAULT_CAPACITY

        newCapacity = @physicalSize + (@physicalSize >> 1) if @physicalSize >= @@DEFAULT_CAPACITY

        temp = Array.new(newCapacity)

        0.upto(@logicalSize - 1) do |i|
          temp[i] = @data[i]
        end

        @data = temp
        @physicalSize = newCapacity
      end

      @logicalSize.downto(index + 1) do |i|
        @data[i] = @data[i - 1]
      end

      @data[index] = element

      @logicalSize += 1
    end

    def remove(index)
      raise "[PANIC - IndexOutOfBounds]" if index.negative? || index >= @logicalSize

      (index + 1).upto(@logicalSize - 1) do |i|
        @data[i - 1] = @data[i]
      end

      @logicalSize -= 1

      @data[@logicalSize] = nil
    end

    def front
      get(0)
    end

    def back
      get(@logicalSize - 1)
    end

    def prepend(element)
      insert(0, element)
    end

    def append(element)
      insert(@logicalSize, element)
    end

    def poll
      remove(0)
    end

    def eject
      remove(@logicalSize - 1)
    end

    def capacity
      @physicalSize
    end

    def shrink
      temp = Array.new(@logicalSize)

      0.upto(@logicalSize - 1) do |i|
        temp[i] = @data[i]
      end

      @data = temp
      @physicalSize = @logicalSize
    end
  end
end

File: python/muse/data_structures/array_list.py
class ArrayList[T]:
    DEFAULT_CAPACITY = 64

    def __init__(self, physicalSize: int = 0):
        self.data: list[T | None] = []
        self.logical_size = 0
        self.physical_size = self.DEFAULT_CAPACITY

        if physicalSize > 1:
            self.physical_size = physicalSize
        self.data = [None] * self.physical_size

    def size(self) -> int:
        return self.logical_size

    def is_empty(self) -> bool:
        return self.logical_size == 0

    def get(self, index: int) -> T:
        if index < 0 or index >= self.logical_size:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        return self.data[index]

    def set(self, index: int, element: T) -> None:
        if index < 0 or index >= self.logical_size:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        self.data[index] = element

    def insert(self, index: int, element: T) -> None:
        if index < 0 or index > self.logical_size:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        if self.logical_size == self.physical_size:
            newCapacity = self.DEFAULT_CAPACITY

            if self.physical_size >= self.DEFAULT_CAPACITY:
                newCapacity = self.physical_size + (self.physical_size >> 1)

            temp: list[T | None] = [None] * newCapacity

            for i in range(self.logical_size):
                temp[i] = self.data[i]

            self.data = temp
            self.physical_size = newCapacity

        for i in range(self.logical_size, index, -1):
            self.data[i] = self.data[i - 1]

        self.data[index] = element

        self.logical_size += 1

    def remove(self, index: int) -> None:
        if index < 0 or index >= self.logical_size:
            raise RuntimeError("[PANIC - IndexOutOfBounds]")

        for i in range(index + 1, self.logical_size):
            self.data[i - 1] = self.data[i]

        self.logical_size -= 1

        self.data[self.logical_size] = None

    def front(self) -> T:
        return self.get(0)

    def back(self) -> T:
        return self.get(self.logical_size - 1)

    def prepend(self, element: T) -> None:
        self.insert(0, element)

    def append(self, element: T) -> None:
        self.insert(self.logical_size, element)

    def poll(self) -> None:
        self.remove(0)

    def eject(self) -> None:
        self.remove(self.logical_size - 1)

    def capacity(self) -> int:
        return self.physical_size

    def shrink(self) -> None:
        temp: list[T | None] = [None] * self.logical_size

        for i in range(self.logical_size):
            temp[i] = self.data[i]

        self.data = temp
        self.physical_size = self.logical_size
